services/src/main/java/org/keycloak/protocol/oidc/AccessTokenIntrospectionProvider.java

```diff
@@ -56,7 +57,11 @@ public Response introspect(String token) {
             if (accessToken != null) {
                 tokenMetadata = JsonSerialization.createObjectNode(accessToken);
                 tokenMetadata.put("client_id", accessToken.getIssuedFor());
-                tokenMetadata.put("username", accessToken.getPreferredUsername());
+
+                UserModel userModel = session.users().getUserById(realm, accessToken.getSubject());
+                if (userModel != null) {
+                    tokenMetadata.put("username", userModel.getUsername());
```

Comment:
IMO this is not the right solution. An introspection should reflect what's in the incoming access token. The "username" in the access token could have been set deliberately to a different value than the user's username (e.g. an external unique user name). And exactly that value should be provided in the introspection result.

```suggestion                   
                    if (!tokenMetadata.has("username") && tokenMetadata.has("preferred_username")) {
                      tokenMetadata.put("username", tokenMetadata.get(accessToken.getPreferredUsername()));
                    }                                      
```
With that approach we get the correct value from the access token and as a fallback have compatibility for those who are relying on the current behavior.
BTW: "client_id" should be treated the same way.

Reply:
"tokenMetadata.put("username", tokenMetadata.get(accessToken.getPreferredUsername()))" do you mean "tokenMetadata.put("username", tokenMetadata.get("preferred_username"))" or "tokenMetadata.put("username", accessToken.getPreferredUsername())" ?

Reply:
Well, hopefully both are returning the same value :-)
tokenMetadata.get() returns a JsonNode, so I thought taking it from the accessToken.getPreferredUsername() is easier.

BTW: If "preferred_username" is also not set, your approach to assign it from UserModel.getUsername() might be OK, as a final fallback.